### PR TITLE
mpl/gpu: Add thread safety to HIP and ZE memory hooks

### PR DESCRIPTION
**Pull request by raffenet**
## Pull Request Description

Add changes made for CUDA hooks in #6866 to HIP and Level Zero modules in MPL. One additional commit to cleanup state after finalize for all 3 modules.

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [ ] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Review by hzhou — APPROVED** [review]
LGTM

File: src/mpl/configure.ac [hunk 1]
```diff
@@ -1194,6 +1194,7 @@ fi
 
 if test "X${pac_have_hip}" = "Xyes" ; then
     AC_DEFINE([HAVE_HIP],[1],[Define if HIP is available])
+    AC_CHECK_LIB(dl, dlopen, [], AC_MSG_ERROR([dlopen not found.  MPL CUDA support requires libdl.]))
```
Comment (abrooks98) [inline src/mpl/configure.ac]:
```suggestion
    AC_CHECK_LIB(dl, dlopen, [], AC_MSG_ERROR([dlopen not found.  MPL HIP support requires libdl.]))
```

**Comment by raffenet** [comment]
@hzhou I added a commit to fix a build issue with HIP support. Please take another quick look.

**Review by abrooks98 — APPROVED** [review]

**Comment by raffenet** [comment]
Thanks @abrooks98!

**Review by hzhou — APPROVED** [review]

**Pull request merged**